src/main/java/org/myrobotlab/service/ProgramAB.java

```diff
@@ -291,9 +296,35 @@ public Response getResponse(String userName, String text) {
     // EEK! clean up the API!
     invoke("publishResponse", response);
     invoke("publishResponseText", response);
-    invoke("publishText", response.msg);
+    invoke("publishText", HtmlFilter.stripHtml(response.msg));
     info("to: %s - %s", userName, res);
 
+    // get optional bot surname, if set..
+    Predicates preds = getChat(userName, getCurrentBotName()).predicates;
+    String botname = getCurrentBotName();
+    if (!preds.get("botname").equals("unknown") && !preds.get("botname").isEmpty()) {
+      botname = preds.get("botname");
+    }
+
+    // feed conversation list
+
+    conversationHistory.add("<b>" + userName + " : </b>" + text);
+    conversationHistory.add("<i><b>" + botname + " : </b>" + HtmlFilter.stripHtml(response.msg) + "</i>");
+
+    // Queue conversationHistory list to prevent unlimited results..
+    if (conversationHistory.size() > maxConversationHistory)
+
+    {
+      List<String> conversationHistoryQueue = new ArrayList<String>();
+      for (int i = conversationHistory.size() - maxConversationHistory; i < conversationHistory.size(); i++) {

```

Comment:
If I understood this correctly (please correct me if I'm wrong):
We're creating a new list and copying elements to it everytime we make a request to the chatbot (and get an response) (if the conversation is running for 100 interactions).
Sorry, this looks like some (unneeded) overhead.

To start this discussion, some alternative solutions that I can currently think of:
Java has a native type Queue which supports adding in the end and removing in the front.
One List and using remove(0) and add(newelement) - would shift all elements when using an ArrayList!
Or a LinkedList to avoid shifting all elements. Now that I think about it, I think LinkedList implements the Queue interface.
Another solution would be a ring buffer, but that would require to keep track of where it starts and ends - more complex in the guis, might not be the best idea.

Reply:
Oh sometime I like to reinvent the wheel ahaha like it. Tks for info I think it is beter to use a native type Queue

Reply:
Have a look at this:  https://stackoverflow.com/questions/5498865/size-limited-queue-that-holds-last-n-elements-in-java/14322473 
we should be using a fixed length queue to maintain conversation history per session.

Reply:
Also.. I want to point out that the "Chat" object in the programab session already has history on it!  Don't re-invent the wheel here!